File: Day0x01/go/main.go
package main

import (
	"bufio"
	"cmp"
	"container/heap"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// Abs computes the absolute value of an integer.
func Abs(x int) int {
	if x >= 0 {
		return x
	}

	return -1 * x
}

// MinHeap implements a min heap data structure.
type MinHeap[T cmp.Ordered] []T

// Len computes the length of a min heap.
func (h MinHeap[T]) Len() int {
	return len(h)
}

// Less compares two elements of a min heap by index and returns
// whether the first one is less than the second.
func (h MinHeap[T]) Less(i, j int) bool {
	return h[i] < h[j]
}

// Swap swaps two elements of a min heap by index.
func (h MinHeap[T]) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

// Push adds a new element to a min heap.
func (h *MinHeap[T]) Push(x any) {
	*h = append(*h, x.(T))
}

// Pop removes and returns the minimum value from a min heap.
func (h *MinHeap[T]) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]

	return x
}

// ParseIDs parses left and right IDs from a single line.
func ParseIDs(s string) (int, int, error) {
	split := strings.Fields(s)
	if len(split) != 2 {
		return 0, 0, fmt.Errorf("expected two values in \"%s\", found %d", s, len(split))
	}

	leftID, err := strconv.Atoi(split[0])
	if err != nil {
		return 0, 0, fmt.Errorf("strconv.Atoi \"%s\": %w", split[0], err)
	}

	rightID, err := strconv.Atoi(split[1])
	if err != nil {
		return 0, 0, fmt.Errorf("strconv.Atoi \"%s\": %w", split[1], err)
	}

	return leftID, rightID, nil
}

func main() {
	filePath := "../lists.txt"
	f, err := os.Open(filePath)
	if err != nil {
		log.Fatalf("ERROR - os.Open: %v\n", err)
		return
	}

	leftHeap := &MinHeap[int]{}
	heap.Init(leftHeap)
	rightHeap := &MinHeap[int]{}
	heap.Init(rightHeap)

	leftCounts := make(map[int]int)
	rightCounts := make(map[int]int)

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		leftID, rightID, err := ParseIDs(line)
		if err != nil {
			log.Fatalf("ERROR - ParseIDs: %v\n", err)
			return
		}

		heap.Push(leftHeap, leftID)
		heap.Push(rightHeap, rightID)

		leftCount, ok := leftCounts[leftID]
		if !ok {
			leftCount = 0
		}
		leftCounts[leftID] = leftCount + 1

		rightCount, ok := rightCounts[rightID]
		if !ok {
			rightCount = 0
		}
		rightCounts[rightID] = rightCount + 1
	}

	totalDistance := 0
	for leftHeap.Len() > 0 && rightHeap.Len() > 0 {
		leftID, rightID := heap.Pop(leftHeap).(int), heap.Pop(rightHeap).(int)
		totalDistance += Abs(leftID - rightID)
	}

	similarityScore := 0
	for leftID, leftCount := range leftCounts {
		rightCount, ok := rightCounts[leftID]
		if !ok {
			rightCount = 0
		}

		similarityScore += leftID * leftCount * rightCount
	}

	fmt.Printf("Part 1: %d\n", totalDistance)
	fmt.Printf("Part 2: %d\n", similarityScore)
}

File: Day0x01/rust/src/main.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::collections::HashMap;
use std::fs;
use std::str::SplitWhitespace;

fn read_file(file_path: &str) -> String {
    return fs::read_to_string(file_path).expect("Should have been able to read file");
}

fn parse_line(line: &str) -> (u32, u32) {
    let mut line_split: SplitWhitespace = line.split_whitespace();
    let left_id: u32 = line_split
        .next()
        .expect("Should have been able to read left id")
        .parse::<u32>()
        .expect("Should have been able to parse left id to u32");
    let right_id: u32 = line_split
        .next()
        .expect("Should have been able to read right id")
        .parse::<u32>()
        .expect("Should have been able to parse right id to u32");

    return (left_id, right_id);
}

fn abs_diff(n1: u32, n2: u32) -> u32 {
    if n1 > n2 {
        return n1 - n2;
    }

    return n2 - n1;
}

fn main() {
    let file_path: &str = "../lists.txt";
    let file_contents: String = read_file(file_path);

    let mut left_heap: BinaryHeap<Reverse<u32>> = BinaryHeap::new();
    let mut right_heap: BinaryHeap<Reverse<u32>> = BinaryHeap::new();
    let mut left_counts: HashMap<u32, u32> = HashMap::new();
    let mut right_counts: HashMap<u32, u32> = HashMap::new();

    for line in file_contents.lines() {
        let (left_id, right_id) = parse_line(line);
        left_heap.push(Reverse(left_id));
        right_heap.push(Reverse(right_id));
    }

    let mut total_distance: u32 = 0;
    while let (Some(Reverse(left_id)), Some(Reverse(right_id))) =
        (left_heap.pop(), right_heap.pop())
    {
        total_distance += abs_diff(left_id, right_id);

        left_counts
            .entry(left_id)
            .and_modify(|count| *count += 1)
            .or_insert(1);
        right_counts
            .entry(right_id)
            .and_modify(|count| *count += 1)
            .or_insert(1);
    }

    let mut similarity_score: u32 = 0;
    for (left_id, left_count) in &left_counts {
        let right_count = right_counts.get(left_id).unwrap_or(&0);
        similarity_score += left_id * left_count * right_count;
    }

    println!("Part 1: {total_distance}");
    println!("Part 2: {similarity_score}");
}
